lib/github-pages/configuration.rb

```diff
@@ -0,0 +1,118 @@
+    #
+    # The following values are also overridden by GitHub Pages, but are not
+    # overridden locally, for practical purposes:
+    # * source
+    # * destination
+    # * jailed
+    # * verbose
+    # * incremental
+    # * GH_ENV
+    OVERRIDES = {
+      "lsi"         => false,
+      "safe"        => true,
+      "plugins"     => SecureRandom.hex,
+      "whitelist"   => PLUGIN_WHITELIST,
+      "highlighter" => "rouge",
+      "kramdown"    => {
+        "template"          => '',
+        'math_engine'       => 'mathajx',
+        'syntax_highligher' => 'rouge'
+      },
+      "gist"        => {
+        "noscript"  => false
+      }
+    }.freeze
+
+    # Options which should be honored *locally* for practical purposes, however
+    # these options are not honored when built by GitHub Pages
+    LOCAL_PASS_THROUGH = %w[
+      config
+      destination
+      source
+      future
+      limit_posts
+      watchforce_polling
+      show_drafts
+      unbpublished
+      quiet
+      verbose
+      incremental
+    ].freeze
+
+    # Given a site, determines the effective configuration by building a user
+    # configuration sandwhich with our overrides overriding the user's specified
+    # values which themselves override our defaults.
+    #
+    # Returns the effective Configuration
+    #
+    # Note: this is a highly modified version of Jekyll#configuration
+    def self.effective_config(site)
+      # Jekyll defaults < GitHub Pages defaults
+      defaults = Jekyll::Utils.deep_merge_hashes(Jekyll::Configuration::DEFAULTS, DEFAULTS)
+
+      # defaults < the site's existing source and destination
+      # so that Jekyll can find the user's config, honor command line flags, etc.
+      passthrough = {}
+      LOCAL_PASS_THROUGH.each { |key| passthrough[key] = site.config[key] }
+      defaults = Jekyll::Utils.deep_merge_hashes(defaults, passthrough)
+
+      # defaults < _config.yml < OVERRIDES
+      config   = Jekyll::Configuration[defaults]
+      override = Jekyll::Configuration[OVERRIDES].stringify_keys
+      config   = config.read_config_files(config.config_files(override))
+      config   = Jekyll::Utils.deep_merge_hashes(config, override).stringify_keys
+
+      # Jekyll's native deep_merge_hashes doesn't merge arrays.
+      # Include default Gems, even if not requested, to avoid breaking pre 3.0 sites
+      config["gems"] = config["gems"].concat(DEFAULT_PLUGINS).uniq
+
+      config
+    end
+
+    # Set the site's configuration  Implemented as an `after_reset` hook.
+    def self.set(site)
+      config = Configuration.effective_config(site)
+
+      # Write the final config to the site object, noting that some values may
+      # have already been set as instancee variables when initialized
+      site.instance_variable_set '@config', config
+      config.keys.each do |key|
+        site.public_send("#{key}=", config[key]) if site.respond_to?("#{key}=")
+      end
```

Comment:
What about [the instance variables set when the site is initialized](https://github.com/jekyll/jekyll/blob/master/lib/jekyll/site.rb#L22-L25)?